common/src/main/java/com/microsoft/identity/common/crypto/AndroidWrappedKeyLoader.java

```diff
@@ -247,17 +257,76 @@ private void saveSecretKeyToStorage(@NonNull final SecretKey unencryptedKey) thr
          * stomping/overwriting one another's keypair.
          */
         KeyPair keyPair = AndroidKeyStoreUtil.readKey(mAlias);
-        if(keyPair == null){
+        if (keyPair == null) {
             Logger.info(methodTag, "No existing keypair. Generating a new one.");
-            keyPair = AndroidKeyStoreUtil.generateKeyPair(
-                    WRAP_KEY_ALGORITHM,
-                    getSpecForKeyStoreKey(mContext, mAlias));
+            final Span span = OTelUtility.createSpan(SpanName.KeyPairGeneration.name());
+            final long keypairGenStartTime = System.currentTimeMillis();
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P && CommonFlightsManager.INSTANCE.getFlightsProvider().isFlightEnabled(CommonFlight.ENABLE_NEW_KEY_GEN_SPEC_FOR_WRAP)) {
+                try (Scope scope = SpanExtension.makeCurrentSpan(span)) {
+                    keyPair = attemptKeyPairGeneration(mAlias, true, keypairGenStartTime);
+                    Logger.info(methodTag, "Successfully generated keypair with new KeyPairGeneratorSpec with wrap purpose.");
+                    Span.current().setAttribute(AttributeName.key_pair_gen_successful_method.name(), "new_key_gen_spec_with_wrap");
+                    span.setStatus(StatusCode.OK);
+                } catch (final ProviderException e) {
+                    if ("SecureKeyImportUnavailableException".equals(e.getClass().getSimpleName())) {
+                        Logger.warn(methodTag, "Wrap purpose may not be supported. Retrying without wrap.");
+                        try {
+                            keyPair = attemptKeyPairGeneration(mAlias, false, keypairGenStartTime);
+                            Logger.info(methodTag, "Successfully generated keypair with new KeyPairGeneratorSpec without wrap purpose.");
+                            Span.current().setAttribute(AttributeName.key_pair_gen_successful_method.name(), "new_key_gen_spec_without_wrap");
+                            span.setStatus(StatusCode.OK);
+                        } catch (Exception ex) {
```

Comment:
Probably worth capturing the exception class name / error message in telemetry. We don't seem to be doing that